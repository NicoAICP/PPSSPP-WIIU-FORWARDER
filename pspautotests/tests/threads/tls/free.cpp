#include "shared.h"

inline void testFree(const char *title, SceUID tlspl) {
	int result = sceKernelFreeTlspl(tlspl);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: OK ", title);
		schedfTlspl(tlspl);
	} else {
		checkpoint(NULL);
		schedf("%s: Failed (%08x) ", title, result);
		schedfTlspl(tlspl);
	}
}

int threadFunc(SceSize args, void *argp) {
	SceUID uid = *(SceUID *)argp;
	void *temp = sceKernelGetTlsAddr(uid);
	checkpoint("  Thread obtained tls: %08x", temp);
	sceKernelSleepThread();
	checkpoint("  Thread quiting without free");
	return 0;
}

extern "C" int main(int argc, char *argv[]) {
	SceUID tls = sceKernelCreateTlspl("tls", PSP_MEMORY_PARTITION_USER, 0, 0x10, 8, NULL);
	void *temp = sceKernelGetTlsAddr(tls);

	checkpointNext("Objects:");
	testFree("  Normal", tls);
	testFree("  Twice", tls);
	testFree("  NULL", 0);
	testFree("  Invalid", 0xDEADBEEF);
	sceKernelDeleteTlspl(tls);
	testFree("  Deleted", tls);

	tls = sceKernelCreateTlspl("tls", PSP_MEMORY_PARTITION_USER, 0, 0x100, 1, NULL);
	temp = sceKernelGetTlsAddr(tls);
	checkpointNext("Threads:");
	{
		TlsplWaitThread wait1("waiting thread", tls);
		testFree("  With waiting thread", tls);
	}
	{
		TlsplWaitThread wait1("holding thread", tls);
		testFree("  With holding thread", tls);
	}

	checkpointNext("Clearing:");
	temp = sceKernelGetTlsAddr(tls);
	memset(temp, 0xCC, 0x100);
	checkpoint("  Before free: %08x", *(u32 *)temp);
	sceKernelFreeTlspl(tls);
	checkpoint("  After free: %08x", *(u32 *)temp);
	// There's only one block, so we must get the same one.
	memset(temp, 0xCC, 0x100);
	temp = sceKernelGetTlsAddr(tls);
	checkpoint("  After allocate: %08x", *(u32 *)temp);
	sceKernelFreeTlspl(tls);

	checkpointNext("Auto free:");
	SceUID threadID = sceKernelCreateThread("no free", &threadFunc, 0x20, 0x1000, 0, NULL);
	const void *arg[1] = { (void *)tls };
	sceKernelStartThread(threadID, sizeof(arg), arg);
	sceKernelDelayThread(1000);
	checkpoint(NULL);
	schedf("  While thread running: ");
	schedfTlspl(tls);
	sceKernelWakeupThread(threadID);
	sceKernelDelayThread(1000);
	checkpoint(NULL);
	schedf("  After thread exit: ");
	schedfTlspl(tls);

	sceKernelDeleteTlspl(tls);
	return 0;
}
