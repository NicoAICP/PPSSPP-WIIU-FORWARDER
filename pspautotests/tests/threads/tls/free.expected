[x] Objects:
[x]   Normal: OK Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000010,totalBlocks=00000008,freeBlocks=00000008,wait=0)
[x]   Twice: OK Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000010,totalBlocks=00000008,freeBlocks=00000008,wait=0)
[x]   NULL: Failed (800201d0) Tlspl: Failed (00000000)
[x]   Invalid: Failed (800201d0) Tlspl: Failed (DEADBEEF)
[x]   Deleted: Failed (800201d0) Tlspl: Invalid (800201D0)

[x] Threads:
[r]   ** started waiting thread
[x]   With waiting thread: OK Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000100,totalBlocks=00000001,freeBlocks=00000000,wait=0)
[r]   ** waiting thread got result: 09d35b00, received = ptr
[r]   ** stopped waiting thread
[r]   ** holding thread got result: 09d35b00, received = ptr
[r]   ** started holding thread
[x]   With holding thread: OK Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000100,totalBlocks=00000001,freeBlocks=00000000,wait=0)
[r]   ** holding thread delayed: 00000000
[r]   ** holding thread freed: 8002019d
[r]   ** stopped holding thread

[x] Clearing:
[x]   Before free: cccccccc
[x]   After free: 00000000
[x]   After allocate: 00000000

[x] Auto free:
[r]   Thread obtained tls: 09d35b00
[r]   While thread running: Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000100,totalBlocks=00000001,freeBlocks=00000000,wait=0)
[r]   Thread quiting without free
[r]   After thread exit: Tlspl: OK (size=60,name=tls,attr=00000000,index=0,blockSize=00000100,totalBlocks=00000001,freeBlocks=00000001,wait=0)
